jupyter_client/ioloop/manager.py

```diff
@@ -15,6 +16,19 @@
 def as_zmqstream(f):
     def wrapped(self, *args, **kwargs):
         socket = f(self, *args, **kwargs)
+        save_socket_class = None
+        # zmqstreams only support sync sockets
+        if self.context._socket_class is not zmq.Socket:
+            save_socket_class = self.context._socket_class
+            self.context._socket_class = zmq.Socket
+        try:
+            orig_socket = socket
+            socket = f(self, *args, **kwargs)
+            orig_socket.close()
```

Comment:
Why is `orig_socket` created and closed, instead of removed entirely?

Reply:
Didn't we need to create it to determine it's type?

Reply:
No, we only check `self.context._socket_class` which we can do before creating sockets

Reply:
Okay, I removed the throwaway socket